drivers/Aqara/aqara-lock/src/credential_utils.lua

```diff
@@ -0,0 +1,151 @@
+local capabilities = require "st.capabilities"
+local utils = require "st.utils"
+
+local remoteControlStatus = capabilities.remoteControlStatus
+local lockCredentialInfo = capabilities["stse.lockCredentialInfo"]
+
+local credential_utils = {}
+
+local HOST_COUNT = "__host_count"
+
+local function save_data(driver)
+  driver.datastore:save()
+  driver.datastore:commit()
+end
+
+local function update_remote_control_status(driver, device, added)
+  local host_cnt = device:get_field(HOST_COUNT) or 0
+  if added then
+    if host_cnt == 0 then
+      device:emit_event(remoteControlStatus.remoteControlEnabled('true', { visibility = { displayed = false } }))
+    end
+    host_cnt = host_cnt + 1
+  else
+    host_cnt = host_cnt - 1
+
+    if host_cnt == 0 then
+      device:emit_event(remoteControlStatus.remoteControlEnabled('false', { visibility = { displayed = false } }))
+    end
+  end
+
+  device:set_field(HOST_COUNT, host_cnt, {persist = true})
+  save_data(driver)
+end
+
+local function sync_all_credential_info(driver, device, command)
+  for _, credentialinfo in ipairs(command.args.credentialInfo) do
+    if credentialinfo.userType == "host" then
+      update_remote_control_status(driver, device, true)
+    end
+  end
+  device:emit_event(lockCredentialInfo.credentialInfo(utils.deep_copy(command.args.credentialInfo), { visibility = { displayed = false } }))
+  save_data(driver)
+end
+
+local function upsert_credential_info(driver, device, command)
+  if #command.args.credentialInfo == 0 then
+    return
+  end
+
+  local credentialInfoTable = utils.deep_copy(device:get_latest_state("main", lockCredentialInfo.ID, lockCredentialInfo.credentialInfo.NAME, {}))
+
+  for _, credentialinfo in ipairs(command.args.credentialInfo) do
+    local exist = false
+    for i = 1, #credentialInfoTable, 1 do
+      if credentialInfoTable[i].credentialId == credentialinfo.credentialId then
+        credentialInfoTable[i] = utils.deep_copy(credentialinfo)
+        exist = true
+        break
+      end
+    end
+
+    if exist == false then
+      if credentialinfo.userType == "host" then
+        update_remote_control_status(driver, device, true)
+      end
+
+      table.insert(credentialInfoTable, credentialinfo)
+    end
+  end
+
+  device:emit_event(lockCredentialInfo.credentialInfo(utils.deep_copy(credentialInfoTable), { visibility = { displayed = false } }))
```

Comment:
Is there a reason `deep_copy` needs to be called here? Can't the table be passed directly?